src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportStream.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+using System.IO.Pipelines;
+using System.Net.Http;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Represents a base WebTransport stream. Do not use directly as it does not
+/// contain logic for handling data.
+/// </summary>
+internal sealed class WebTransportStream : ConnectionContext, IStreamDirectionFeature
+{
+    private readonly CancellationTokenRegistration _connectionClosedRegistration;
+    private readonly bool _canWrite;
+    private readonly bool _canRead;
+    private readonly DuplexPipe _duplexPipe;
+    private readonly IFeatureCollection _features;
+    private readonly KestrelTrace _log;
+    private IDictionary<object, object?>? _items;
+    private bool _isClosed;
+
+    public readonly long StreamId;
+
+    /// <summary>
+    /// Use StreamId instead of this whenever possible to avoid the cast from long to string
+    /// </summary>
+    public override string ConnectionId { get => StreamId.ToString(NumberFormatInfo.InvariantInfo); set => throw new NotSupportedException(); }
+
+    public override IDuplexPipe Transport { get => _duplexPipe; set => throw new NotSupportedException(); }
+
+    public override IFeatureCollection Features => _features;
+
+    public override IDictionary<object, object?> Items
+    {
+        get => _items ??= new Dictionary<object, object?>();
+        set => _items = value;
+    }
+
+    public bool CanRead => _canRead && !_isClosed;
+
+    public bool CanWrite => _canWrite && !_isClosed;
+
+    internal WebTransportStream(Http3StreamContext context, WebTransportStreamType type)
+    {
+        _canRead = type != WebTransportStreamType.Output;
+        _canWrite = type != WebTransportStreamType.Input;
+        _log = context.ServiceContext.Log;
+
+        var streamIdFeature = context.ConnectionFeatures.GetRequiredFeature<IStreamIdFeature>();
+        StreamId = streamIdFeature!.StreamId;
+
+        _features = context.ConnectionFeatures;
```

Comment:
Is `Connection` here the Quic connection or the QuicStream virtual connection?

Reply:
I am not sure what you mean by `Connection` but the `context` is an `Http3StreamContext` that is created from the `ConnectionContext` that the `_multiplexedContext` returns when you create or accept a Quic stream.